stan-dev/cmdstan
# Bugfix/763 765 makefiles

#### Submisison Checklist


#### Summary:

change make rules so that all necessary dependencies files are generated during model compilation for all models.

bug introduced in 2.20 - added a rule to `make/tests` which did the wrong thing w/r/t dependencies files - it correctly generated dependencies files for the cmdstan test models; but for all other models - compiled by the rules in `make/program` - the rule no longer created `.d` as part of the model compilation step.   

this resulted in bugs reported in https://github.com/stan-dev/cmdstan/issues/763 and https://github.com/stan-dev/cmdstan/issues/765.

#### Intended Effect:

trigger recompilation as needed whenever there's a change to a model of core stan.

#### How to Verify:

use make to build and rebuild programs - need to automate this test

to check that models are recompiled whenever the .hpp file goes away,  given a model `foo.stan`, do:
   + `make foo`
   +  `rm foo.hpp`
   + `make foo` - should trigger rebuild

- to check that models are recompiled whenever core Stan changes, given a model `foo.stan` do:
  + `make foo`
  + edit your favorite file in core Stan, e.g. stan/src/mcmc/var_adaptation.hpp
  + `make foo` - should trigger rebuild of main.o

#### Side Effects:

N/A

#### Documentation:

N/A

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
